lib/evm/operation.ex

```diff
@@ -0,0 +1,340 @@
+
+      iex> EVM.Operation.metadata(0x00)
+      %EVM.Operation.Metadata{id: 0x00, sym: :stop, input_count: 0, output_count: 0, description: "Halts execution", group: :stop_and_arithmetic}
+
+      iex> EVM.Operation.metadata(:add)
+      %EVM.Operation.Metadata{id: 0x01, sym: :add, input_count: 2, output_count: 1, description: "Addition operation", group: :stop_and_arithmetic}
+
+      iex> EVM.Operation.metadata(:push1)
+      %EVM.Operation.Metadata{id: 0x60, sym: :push1, fun: :push_n, args: [1], input_count: 0, output_count: 1, description: "Place 1-byte item on stack", group: :push}
+
+      iex> EVM.Operation.metadata(0xfe)
+      nil
+
+      iex> EVM.Operation.metadata(nil)
+      nil
+  """
+  @spec metadata(operation | opcode) :: EVM.Operation.Metadata.t | nil
+  def metadata(nil), do: nil
+  def metadata(operation) when is_atom(operation) do
+    operation |> encode |> metadata
+  end
+
+  def metadata(opcode) when is_integer(opcode) do
+    Map.get(@opcodes_to_metadata, opcode)
+  end
+
+  @doc """
+  Executes a single operation. This simply does the effects of the operation itself,
+  ignoring the rest of the actions of an operation cycle. This will effect, for instance,
+  the stack, but will not effect the gas, etc.
+
+  ## Examples
+
+      # TODO: How to handle trie state in tests?
+
+      # Add
+      iex> EVM.Operation.run_operation(:add, %{}, %EVM.MachineState{stack: [1, 2]}, %EVM.SubState{}, %EVM.ExecEnv{})
+      {%{}, %EVM.MachineState{stack: [3]}, %EVM.SubState{}, %EVM.ExecEnv{}}
+
+      # Push
+      iex> EVM.Operation.run_operation(:push1, %{}, %EVM.MachineState{stack: [1, 2]}, %EVM.SubState{}, %EVM.ExecEnv{machine_code: <<00, 01>>})
+      {%{}, %EVM.MachineState{stack: [1, 1, 2]}, %EVM.SubState{}, %EVM.ExecEnv{machine_code: <<00, 01>>}}
+
+      # nil
+      iex> EVM.Operation.run_operation(:stop, %{}, %EVM.MachineState{stack: [1, 2]}, %EVM.SubState{}, %EVM.ExecEnv{})
+      {%{}, %EVM.MachineState{stack: [1, 2]}, %EVM.SubState{}, %EVM.ExecEnv{}}
+
+      # Unimplemented
+      iex> EVM.Operation.run_operation(:log0, %{}, %EVM.MachineState{stack: [1, 2]}, %EVM.SubState{}, %EVM.ExecEnv{})
+      {%{}, %EVM.MachineState{stack: []}, %EVM.SubState{}, %EVM.ExecEnv{}}
+  """
+  @spec run_operation(operation, EVM.state, MachineState.t, SubState.t, ExecEnv.t) :: {EVM.state, MachineState.t, SubState.t, ExecEnv.t}
+  def run_operation(operation, state, machine_state, sub_state, exec_env) do
+    {args, updated_stack, updated_machine_state} = operation_args(operation, state, machine_state, sub_state, exec_env)
+
+    apply_to_group_module(operation, args)
+      |> normalize_op_result(updated_stack)
+      |> merge_state(
+        operation,
+        state,
+        updated_machine_state,
+        sub_state,
+        exec_env
+      )
+  end
+
+  defp apply_to_group_module(operation, args) do
```

Comment:
Can you add a type spec?

Reply:
(I've fallen in love with how clear the typespecs make the code, and they sorted out a ton of bugs when I ran dialyzer)

Reply:
Yep! Good call: [Add typespecs](https://github.com/exthereum/evm/pull/3/commits/2f1d1f3552c1770722305d5c0e88d2665173fb25)

I want to get into the habit of doing this! :)

I can imagine static analysis makes things much more reliable.